97.md

```diff
@@ -0,0 +1,251 @@
+relay:  ["ATTACH",<attachment-id>, true, "complete"]
+     //event is published
+```
+
+The protocol flow is designed in such a way that it is possible to post other commands between individual phases. Possible implementation on a relay:
+
+```
+- on ATTACH command
+   - validate the event
+   - store the event in a temporary storage associated with current connection
+
+- on binary message
+   - calculate SHA256 hash of the binary message
+   - find matching attachment in list of attachment of events stored in the temporary storage
+   - store the binary message as new attachment in the temporary storage
+   - check if all attachments have been uploaded
+        - is so, publish the stored event and attachments and delete them from temporary storage
+
+ - on connection closed before completion
+   - destroy temporary storage (no publish)
+   
+```
+
+
+**NOTE** If the relay does not support this NIP, it will not be able to respond to the new command, or it will respond with "NOTICE". If the relay does not respond to the command within a certain time (e.g. 10 seconds), then the client should verify that the relay responds to other commands (for example "REQ") and evaluate the situation as if the relay does not support the functionality. If the relay responds to the command with a "NOTICE" response, the client must assume that the relay does not support the function and try another method
+
+Response - ATTACH
+--------------------------------------
+
+Response for ATTACH is similar as response to EVENT. It uses "OK" response.
+
+```
+relay:  ["OK","<event-id">,true,"continue"]
+```
+
+The fourth argument may contain more detailed information about the operation. 
+* `"continue"` - accepted, continue sending attachments
+* `"duplicate"` - event is duplicate, you can still send attachments.
+
+Errors - ATTACH
+-----------------------------
+When an error occurs after an ATTACH command, the client must not send any binary messages in this case.
+
+Error description is specified as fouth argument
+
+```
+relay:  ["OK","<event-id">,false,"error"]
+```
+
+* `"max_attachment_size: <number>"` - one of the attachments exceeded the maximum size allowed, the relay sends this limit as a number in the error description section
+* `"max_attachment_count: <number>"` - the number of attachments has exceeded the allowed limit. Again, the relay sends this limit as a number in the error description section
+* `"<event-id">,false,"invalid: <description>"` - event is malformed, missing mandatory fields, or event doesn't have attachments
+
+
+Response - binary messages
+--------------------------------------
+
+Relay must respond to each binary message with a ["ATTACH"] response. The response has the same format as "OK"
+
+The fourth argument may contain more detailed information about the operation. 
+
+```
+relay:  ["ATTACH","<attachment-id">,true,"description"]
+```
+
+* `"continue"` - accepted, continue sending attachments
+* `"complete"` - event has been posted
```

Comment:
If a client uploads an event with 3 attachments but uploads only 2, and posts it, what happens? And what happens if the client tries to send the last attachment a few seconds later? Will the relay have deleted the event and attachments already? Will it allow a new attachment in? Will it just remain incomplete forever?

Reply:
This is why there is a requirement for it to have transactional behavior (atomic)

By issuing the command ATTACH you open an transaction on the relay
while the transaction is active, garbage collector cannot delete attachments.
Transaction is automatically committed by posting the last attachment (complete status)
Transaction is automatically rollbacked when you close connection or when you issue another ATTACH command.

In my implementation, I store all attachments in the main storage immediately, however the transaction holds a special object `AttachmentLock`, which serves as reference to an attachment (so GC can see that reference). This lock lives on the transaction until it is committed or rollbacked.

Reply:
I would just not have the transaction scope at all and use a timeout for new files. GC can only delete after 5 minutes since posted. It's the same as a lock but you don't need to expect clients to behave (release the lock)

Reply:
I have always problem with magic constants in a code.

Where I write anything about releasing lock?

Reply:
I don't think you can avoid the magic constant. You will need a time-out because the client might never finish uploading everything (it might/will crash)

As a client, the specific timeout a relay uses doesn't matter. I have to assume a broken database nevertheless.